nixos/modules/services/home-automation/home-assistant.nix

```diff
@@ -42,33 +35,72 @@ let
   #   platform = "mqtt";
   #   ...
   # } ];
-  useComponentPlatform = component: elem component (usedPlatforms cfg.config);
+  usedPlatforms = config:
+    if isAttrs config then
+      optional (config ? platform) config.platform
+      ++ concatMap usedPlatforms (attrValues config)
+    else if isList config then
+      concatMap usedPlatforms config
+    else [ ];
 
-  useExplicitComponent = component: elem component explicitComponents;
+  useComponentPlatform = component: elem component (usedPlatforms cfg.config);
 
-  # Returns whether component is used in config or explicitly passed into package
+  # Returns whether component is used in config, explicitly passed into package or
+  # configured in the module.
   useComponent = component:
     hasAttrByPath (splitString "." component) cfg.config
     || useComponentPlatform component
-    || useExplicitComponent component;
+    || useExplicitComponent component
+    || builtins.elem component cfg.extraComponents;
 
-  # List of components used in config
+  # Final list of components passed into the package to include required dependencies
   extraComponents = filter useComponent availableComponents;
+  extraComponentsFile = format.generate "extraComponents.yaml" extraComponents;
 
-  package = if (cfg.autoExtraComponents && cfg.config != null)
-    then (cfg.package.override { inherit extraComponents; })
+  # Final list of extra packages passed into the package, that also includes custom
+  # component dependencies
+  extraPackages = cfg.extraPackages
+    ++ map (component: component.dependencies or []) cfg.customComponents;
+
+  package = if (extraComponents != [] || extraPackages != [])
+    then (cfg.package.override {
+      inherit extraComponents;
+      extraPackages = ps: with ps; cfg.extraPackages;
+    })
     else cfg.package;
 
-  # If you are changing this, please update the description in applyDefaultConfig
-  defaultConfig = {
-    homeassistant.time_zone = config.time.timeZone;
-    http.server_port = cfg.port;
-  } // optionalAttrs (cfg.lovelaceConfig != null) {
-    lovelace.mode = "yaml";
+  # Create a directory that holds all custom components
+  customComponentsDir = pkgs.buildEnv {
+    name = "home-assistant-custom-components";
+    paths = cfg.customComponents;
+  };
+
+  # Create a directory that holds all lovelace modules
+  customLovelaceModulesDir = pkgs.buildEnv {
+    name = "home-assistant-custom-lovelace-modules";
+    paths = cfg.customLovelaceModules;
+  };
+
+  # Create parts of the lovelace config that references lovelave modules as resources
+  customLovelaceModulesResources = {
+    lovelace.resources = map (card: {
+      url = "/local/nixos-lovelace-modules/${card.pname}.js?${card.version}";
```

Comment:
Wondering if the lovelace module packages should set something like `passthru.entrypoint` that we reference here instead of the `pname`.